source/extensions/network/dns_resolver/cares/dns_impl.h

```diff
@@ -113,5 +115,36 @@ class DnsResolverImpl : public DnsResolver, protected Logger::Loggable<Logger::I
   const absl::optional<std::string> resolvers_csv_;
 };
 
+class CaresDnsResolverFactoryImpl : public DnsResolverFactory {
+public:
+  std::string name() const override { return cares_dns_resolver; }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return ProtobufTypes::MessagePtr{
+        new envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig()};
+  }
+
+  DnsResolverSharedPtr createDnsResolverCb(
+      Event::Dispatcher& dispatcher, const Api::Api&,
+      const envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) override {
+    envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig cares;
+    envoy::config::core::v3::DnsResolverOptions dns_resolver_options;
+    std::vector<Network::Address::InstanceConstSharedPtr> resolvers;
+
+    // Only cares DNS factory will call into this function.
+    // Directly unpack the typed config to a cares object.
+    Envoy::MessageUtil::unpackTo(typed_dns_resolver_config.typed_config(), cares);
+    dns_resolver_options.MergeFrom(cares.dns_resolver_options());
+    if (!cares.resolvers().empty()) {
+      const auto& resolver_addrs = cares.resolvers();
+      resolvers.reserve(resolver_addrs.size());
+      for (const auto& resolver_addr : resolver_addrs) {
+        resolvers.push_back(Network::Address::resolveProtoAddress(resolver_addr));
```

Comment:
This logic is inherited from existing code, for example : https://github.com/envoyproxy/envoy/blob/16879db8d6d3e859e169807e7b9510ba4b30501f/source/server/server.cc#L593. The existing code did not whether check the address is valid, maybe because the origination of the resolvers() are from configuration with the field type: config.core.v3.Address, which only can have:  SocketAddress, Pipe, or EnvoyInternalAddress, i.e, all valid.

There are test cases to cover this code in dns_impl_test.cc:  TEST: SupportCustomResolvers and TEST:SupportsMultipleCustomResolversAndDnsOptions.